druid/src/widget/svg.rs

```diff
@@ -192,91 +197,160 @@ impl FromStr for SvgData {
+                    x1,
+                    y1,
+                    x2,
+                    y2,
+                    x,
+                    y,
+                } => {
+                    path.curve_to((x1, y1), (x2, y2), (x, y));
+                }
+                usvg::PathSegment::ClosePath => {
+                    path.close_path();
+                }
             }
-            usvg::PathSegment::CurveTo {
-                x1,
-                y1,
-                x2,
-                y2,
-                x,
-                y,
-            } => {
-                path.curve_to((x1, y1), (x2, y2), (x, y));
+        }
+
+        path.apply_affine(self.offset_matrix * transform_to_affine(p.transform));
+
+        match &p.fill {
+            Some(fill) => {
+                let brush = self.brush_from_usvg(&fill.paint, fill.opacity, ctx);
+                ctx.fill(path.clone(), &*brush);
             }
-            usvg::PathSegment::ClosePath => {
-                path.close_path();
+            None => {}
+        }
+
+        match &p.stroke {
+            Some(stroke) => {
+                let brush = self.brush_from_usvg(&stroke.paint, stroke.opacity, ctx);
+                ctx.stroke(path, &*brush, stroke.width.value());
             }
+            None => {}
         }
     }
 
-    path.apply_affine(offset_matrix * transform_to_affine(p.transform));
-
-    match &p.fill {
-        Some(fill) => {
-            let brush = color_from_usvg(&fill.paint, fill.opacity);
-            ctx.fill(path.clone(), &brush);
-        }
-        None => {}
+    fn linear_gradient_def(&mut self, lg: &usvg::LinearGradient, ctx: &mut PaintCtx) {
+        // Get start and stop of gradient and transform them to image space (TODO check we need to
+        // apply offset matrix)
+        let start = self.offset_matrix * Point::new(lg.x1, lg.y1);
+        let end = self.offset_matrix * Point::new(lg.x2, lg.y2);
+        let stops: Vec<_> = lg
+            .base
+            .stops
+            .iter()
+            .map(|stop| GradientStop {
+                pos: stop.offset.value() as f32,
+                color: color_from_svg(stop.color, stop.opacity),
+            })
+            .collect();
+
+        // TODO error handling
+        let gradient = FixedLinearGradient { start, end, stops };
+        println!("{} => {:?}", lg.id, gradient);
+        let gradient = ctx.gradient(gradient).unwrap();
+        self.defs.add_def(lg.id.clone(), gradient);
     }
 
-    match &p.stroke {
-        Some(stroke) => {
-            let brush = color_from_usvg(&stroke.paint, stroke.opacity);
-            ctx.stroke(path, &brush, stroke.width.value());
+    fn brush_from_usvg(
+        &self,
+        paint: &usvg::Paint,
+        opacity: usvg::Opacity,
+        ctx: &mut PaintCtx,
+    ) -> Rc<piet::Brush> {
+        match paint {
+            usvg::Paint::Color(c) => {
+                // TODO I'm going to assume here that not retaining colors is OK.
```

Comment:
Are color brushes always cheap to create?

Reply:
This is not a guarantee made by piet, but ideally I think that the caching of brushes, especially color brushes, should be something that is handled at the piet level. Gradient brushes are definitely not necessarily cheap to create.